File: feature_tests/java/somelib/src/main/java/dev/diplomattest/somelib/RefList.java
package dev.diplomattest.somelib;

import dev.diplomattest.somelib.ntv.*;


import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.lang.foreign.SegmentAllocator;
import java.lang.ref.Cleaner;
import java.util.List;
import static java.lang.foreign.ValueLayout.*;
import java.nio.charset.StandardCharsets;
import java.util.stream.Stream;

public class RefList {

    MemorySegment internal;
    Cleaner.Cleanable cleanable;
    SegmentAllocator arena;

    List<Object> selfEdges = List.of();
    List<Object> aEdges = List.of();
    

    static class RefListCleaner implements Runnable {

        MemorySegment segment;
        RefListCleaner(MemorySegment segment) {
            this.segment = segment;
        }

        public void run() {
            somelib_h.RefList_destroy(this.segment);
        }
    }

    RefList() {}
    RefList(MemorySegment handle, List<Object> selfEdges, List<Object> aEdges) {
        this.internal = handle;
        this.selfEdges = selfEdges;
        this.aEdges = aEdges;
        

    }
    
    public static RefList node(RefListParameter data) {
        
        var dataNative = data.internal;
        var nativeVal = somelib_h.RefList_node(dataNative);
        
        List<Object> selfEdges = List.of();
        
        
        
        List<Object> bEdges = List.of(data);
        var returnVal = new RefList(nativeVal, selfEdges, bEdges);
        return returnVal;
                
    }
    
    
}

File: feature_tests/kotlin/somelib/src/main/kotlin/dev/diplomattest/somelib/RefList.kt
package dev.diplomattest.somelib;
import com.sun.jna.Callback
import com.sun.jna.Library
import com.sun.jna.Native
import com.sun.jna.Pointer
import com.sun.jna.Structure

internal interface RefListLib: Library {
    fun RefList_destroy(handle: Pointer)
    fun RefList_node(data: Pointer): Pointer
}

class RefList internal constructor (
    internal val handle: Pointer,
    // These ensure that anything that is borrowed is kept alive and not cleaned
    // up by the garbage collector.
    internal val selfEdges: List<Any>,
    internal val aEdges: List<Any?>,
)  {

    internal class RefListCleaner(val handle: Pointer, val lib: RefListLib) : Runnable {
        override fun run() {
            lib.RefList_destroy(handle)
        }
    }

    companion object {
        internal val libClass: Class<RefListLib> = RefListLib::class.java
        internal val lib: RefListLib = Native.load("diplomat_feature_tests", libClass)
        @JvmStatic
        
        fun node(data: RefListParameter): RefList {
            // This lifetime edge depends on lifetimes: 'b
            val bEdges: MutableList<Any> = mutableListOf(data);
            
            val returnVal = lib.RefList_node(data.handle);
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = RefList(handle, selfEdges, bEdges)
            CLEANER.register(returnOpaque, RefList.RefListCleaner(handle, RefList.lib));
            return returnOpaque
        }
    }

}